optimum/habana/transformers/generation/utils.py

```diff
@@ -189,6 +192,116 @@ def _update_model_kwargs_for_generation(
+        batch_size: int,
+        pad_token_id: int,
+        device: str,
+        dtype: str = torch.int64,
+    ) -> torch.Tensor:
+        x = torch.zeros((batch_size, max_steps), device=device, dtype=dtype)
+        return x.index_fill(1, torch.tensor([0]), 1)  # First the position with pad_token_id
+
+    def _prepare_past_key_values(
+        self,
+        max_steps: int,  # current stopping criteria
+        batch_size: int,
+        device: str,
+    ) -> Tuple[torch.LongTensor, torch.LongTensor, Optional[torch.Tensor], Optional[torch.Tensor]]:
+        n_heads = self.config.decoder_attention_heads
+        embed_size_per_head = self.config.d_model // n_heads
+        key_states_ = torch.zeros(
+            (self.config.decoder_layers, batch_size, n_heads, max_steps, embed_size_per_head), device=device
+        )
+        value_states_ = torch.zeros(
+            (self.config.decoder_layers, batch_size, n_heads, max_steps, embed_size_per_head), device=device
+        )
+        past_key_values = tuple(
+            (key_states_[i, :, :, :, :], value_states_[i, :, :, :, :], None, None)
+            for i in range(self.config.decoder_layers)
+        )
+        return past_key_values
+
+    def _prepare_decoder_input_ids_for_generation(
+        self,
+        batch_size: int,
+        model_input_name: str,
+        model_kwargs: Dict[str, torch.Tensor],
+        decoder_start_token_id: int = None,
+        bos_token_id: int = None,
+        device: torch.device = None,
+        max_length: int = 1,
+    ) -> Tuple[torch.LongTensor, Dict[str, torch.Tensor]]:
+        """Prepares `decoder_input_ids` for generation with encoder-decoder models"""
+        # 1. Check whether the user has defined `decoder_input_ids` manually. To facilitate in terms of input naming,
+        # we also allow the user to pass it under `input_ids`, if the encoder does not use it as the main input.
+
+        if model_kwargs is not None and "decoder_input_ids" in model_kwargs:
+            decoder_input_ids = model_kwargs.pop("decoder_input_ids")
+        elif "input_ids" in model_kwargs and model_input_name != "input_ids":
+            decoder_input_ids = model_kwargs.pop("input_ids")
+        else:
+            decoder_input_ids = None
+
+        token_idx = model_kwargs.get("token_idx", None)
+
+        # 2. Encoder-decoder models expect the `decoder_input_ids` to start with a special token. Let's ensure that.
+        decoder_start_token_id = self._get_decoder_start_token_id(decoder_start_token_id, bos_token_id)
+        if device is None:
+            device = self.device
+        if token_idx is None:
+            decoder_input_ids_start = (
+                torch.ones((batch_size, 1), dtype=torch.long, device=device) * decoder_start_token_id
+            )
+        else:
+            decoder_input_ids_start = (
+                torch.ones((batch_size, max_length), dtype=torch.long, device=device) * decoder_start_token_id
```

Comment:
```suggestion
                torch.ones((batch_size, self.generation_config.max_length), dtype=torch.long, device=device) * decoder_start_token_id
```
This should work, and that way we can remove the extra `max_length` argument to this method.